common/core/gpudevice.cpp

```diff
@@ -663,14 +668,11 @@ void GpuDevice::HandleRoutine() {
   // 2) Another app is having control of display and we
   //    we need to take control.
   // TODO: Add splash screen support.
-  lock_fd_ = open("/vendor/hwc.lock", O_RDONLY);
-  if (lock_fd_ == -1) {
-    thread_sync_lock_.unlock();
-    return;
-  }
 
-  thread_sync_lock_.unlock();
-  display_manager_->IgnoreUpdates();
+  if (flock(lock_fd_, LOCK_EX|LOCK_NB) != 0) {
+    ETRACE("Getting the exlusive file lock on /vendor/hwc.lock failed.");
+    display_manager_->IgnoreUpdates();
+  }
```

Comment:
If exclusive lock grabbed immediately, then there is no other process hold the lock. So no need to call IgnoreUpdates method. The else statement print some version messages regarding to grabbed the exclusive lock?

Reply:
Your log is confusing "Getting the exlusive file lock on /vendor/hwc.lock failed" and u call ignore updates. Also, flock( below is that needed ?

Reply:
First non-blocking flock on lock_fd_ is let thread know if there is a process has been holding the exclusive lock. Since it's non-blocking, it returns immediately if the lock is really hold by another process, then it has a chance to call IgnoreUpdates to avoid HWC rendering content into DRM

Then it continues to try grab exclusive lock with blocking mode and fall into sleep wait until another process release the lock. Once the call returns, it call the ForceUpdate to let HWC rendering content to DRM.

May be we could show message like "Another process is holding /vendor/hwc.lock, wait until it release the lock." when first flock non-blocking returns with error.

Reply:
No need to grab again if first non-blocking call successfully because no other process holding the lock. If first non-blocking call returns with error, that means another process is holding lock and second exclusive lock is required. Second blocking exclusive lock is to ensure HWC GpuDevice thread fall into sleep wait and resumed after the lock released from another process.  After got the exclusive lock, it's time to call ForceUpdate to take over the DRM ownership.

Reply:
I understand that but we dont have the checks, we are doing it always now.

Reply:
if (flock(lock_fd_, LOCK_EX|LOCK_NB) != 0) {
 display_manager_->IgnoreUpdates();
  if (flock(lock_fd_, LOCK_EX) != 0)
   ETRACE("Failed to wait on hwc lock.");
  } else {
ETRACE("Getting the exlusive file lock on /vendor/hwc.lock failed.");
close(lock_fd_);
lock_fd_ = -1;
}

?

Reply:
if (flock(lock_fd_, LOCK_EX|LOCK_NB) != 0) {
display_manager_->IgnoreUpdates();
if (flock(lock_fd_, LOCK_EX) != 0)
ETRACE("Failed to wait on hwc lock.");
} else {
ETRACE("Getting the exlusive file lock on /vendor/hwc.lock failed.");
}
close(lock_fd_);
lock_fd_ = -1;

Yes, code pieces regarding to lock works for me and made some little changes on lock_fd_ close up. lock_fd_ has to be closed and set to -1. Otherwise, next HandleWait will not fall into endless sleep :-)

Works for you?

Reply:
ya